### PR TITLE
PLG-731: Adapt the calculation of KIs for product models

### PR DESCRIPTION
**Pull request by BryanS29-akeneo**
**Description (for Contributor and Core Developer)**

To adapt the calculation of key indicators about image and spelling, we had to separate products and product models with dependency injection and add some tests.

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\Query\ProductEvaluation\GetEvaluationRatesByProductsAndCriterionQueryInterface;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\CriterionCode;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductId;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductIdCollection;
+use Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Transformation\TransformCriterionEvaluationResultCodes;
+use Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Transformation\TransformCriterionEvaluationResultIds;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
```
Comment (LaurentPetard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php:16]:
```suggestion
 * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class GetEvaluationRatesByProductModelsAndCriterionQuery implements GetEvaluationRatesByProductsAndCriterionQueryInterface
+{
+    public function __construct(private Connection $dbConnection, private TransformCriterionEvaluationResultIds $transformCriterionEvaluationResultIds)
+    {
+    }
```
Comment (LaurentPetard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php:23]:
```suggestion
    public function __construct(
        private Connection $dbConnection, 
        private TransformCriterionEvaluationResultIds $transformCriterionEvaluationResultIds
    ) {
    }
```

Comment (LaurentPetard) [reply src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php:23]: it's more readable with line breaks after each property

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    private function formatEvaluationRates(array $evaluationResult): array
+    {
+        if (!isset($evaluationResult['rates'])) {
+            return [];
+        }
+
+        $rates = json_decode($evaluationResult['rates'], true, 512, JSON_THROW_ON_ERROR);
+        $rates = $this->transformCriterionEvaluationResultIds->transformToCodes(([TransformCriterionEvaluationResultCodes::PROPERTIES_ID['rates'] => $rates]));
```
Comment (LaurentPetard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php:61]: It seems that there is useless `(` and `)`  (same for `GetEvaluationRatesByProductsAndCriterionQuery`)

File: src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Integration/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQueryIntegration.php [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\CriterionEvaluationStatus;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\LocaleCode;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductId;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductIdCollection;
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\Rate;
+use Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Query\ProductEvaluation\GetEvaluationRatesByProductModelsAndCriterionQuery;
+use Akeneo\Test\Pim\Automation\DataQualityInsights\Integration\DataQualityInsightsTestCase;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
```
Comment (LaurentPetard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Integration/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQueryIntegration.php:20]:
```suggestion
 * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
```

**Review by LaurentPetard — APPROVED** [review]

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductIdCollection;
+use Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Transformation\TransformCriterionEvaluationResultCodes;
+use Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Transformation\TransformCriterionEvaluationResultIds;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class GetEvaluationRatesByProductModelsAndCriterionQuery implements GetEvaluationRatesByProductsAndCriterionQueryInterface
```
[the remark this answers is not in the document]
Comment (BryanS29-akeneo) [reply src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php:19]: Class name of queries are usually long in this directory `src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation`, I think it's to easily recognize what the query do.

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+{
+    public function __construct(
+        private Connection $dbConnection,
+        private TransformCriterionEvaluationResultIds $transformCriterionEvaluationResultIds
+    ) {
+    }
+
+    public function execute(ProductIdCollection $productIdCollection, CriterionCode $criterionCode): array
```
[the remark this answers is not in the document]
Comment (BryanS29-akeneo) [reply src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php:27]: The result it's a multidimensional array to complex to type in phpdoc : https://github.com/akeneo/pim-community-dev/blob/d075911f5f1822f333edde00cdf1d74f4ba4988a/src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetEvaluationRatesByProductModelsAndCriterionQuery.php#L48-L66

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Symfony/Resources/config/services.yml [hunk 7]
```diff
@@ -188,10 +188,17 @@ services:
...
         arguments:
             - '@Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Query\Structure\CachedGetLocalesByChannelQuery'
-            - !tagged akeneo.pim.automation.data_quality_insights.compute_key_indicator_query
+            - !tagged akeneo.pim.automation.data_quality_insights.compute_product_key_indicator_query
+
+    akeneo.pim.automation.data_quality_insights.compute_product_models_key_indicators:
+        class: Akeneo\Pim\Automation\DataQualityInsights\Application\ComputeProductsKeyIndicators
+        arguments:
+            - '@Akeneo\Pim\Automation\DataQualityInsights\Infrastructure\Persistence\Query\Structure\CachedGetLocalesByChannelQuery'
+            - !tagged akeneo.pim.automation.data_quality_insights.compute_product_model_key_indicator_query
```
[the remark this answers is not in the document]
Comment (BryanS29-akeneo) [reply src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Symfony/Resources/config/services.yml:201]: I don't know exactly.. it's interesting, you can ask in our channel

[the remark this answers is not in the document]
Comment (olivierpontier) [reply src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Symfony/Resources/config/services.yml:201]: It is a native shortcut used during the compilation phase of the Kernel. Symfony will aggregate all services tagged with the `akeneo.pim.automation.data_quality_insights.compute_product_model_key_indicator_query` tag here and inject this list as an iterator. Doc: https://symfony.com/doc/current/service_container/tags.html

Before this syntax, we used to define the list services manually or create a dedicated CompilerPass to handle it automatically

**Pull request merged**